### PR TITLE
Govstack api spp_base_gis

### LINKED ISSUE #344: Development of spp_base_gis module
Objective:
To have our own GIS engine where we can plot vectors (Point, Line, Polygon) on a map by utilizing the maptiler JS and to have our own customized view where we can see on the map all of the vectors of the record.

Features:
- base module for GIS where a customized field will be set that will automatically render a Map whenever the field is called on the UI.
- a new customized view that will collect all of the vectors of all of the records of a model and show all of them in the map
- Able to customize the vectors
- Able to customize the UI of the map in the new customized view
- Able to add a raster layer in the new customized view
- Able to search for a country and city
- Able to click the vectors to show the details of the record
- Able to get the currect location of the device
- Able to search a specific record

### PR DESCRIPTION
**Pull request by reichie020212**
- Added API endpoint for GIS locational query
- Added API endpoint for GIS attribute query
- Added Test Cases
- Currently supported spatial relation: within, intersects, contains, touches, crosses, equals, disjoint, covers, coveredby
- Able to search them via odoo's ORM using below operators
-> gis_intersects, gis_contains, gis_touches, gis_crosses, gis_equals, gis_disjoint, gis_covers, gis_coveredby
- here is a sample odoo search orm
-> `env["res.partner"].search([("geo_polygon_field", "gis_intersects", "POINT(1 2)")])`
- value in odoo search orm can support these types: JSON (in geojson format), dictionary (in geojson format), WKT (Well-Known Text), and Geometry Shape
- Sample JSON
-> `'{"type": "Point", "coordinates": [124.98706532650226, 7.931953944598931]}'`
- Sample dictionary
-> `{'type': 'Point', 'coordinates': [124.98706532650226, 7.931953944598931]}`
- Sample WKT
-> `'POINT (124.98706532650226 7.931953944598931)'`
- Sample Shape
```
-> from shapely.geometry import Point
-> Point(124.98706532650226, 7.931953944598931)
```

other samples

```
env["res.partner"].search([("geo_polygon_field", "gis_intersects", '{"type": "Point", "coordinates": [124.98706532650226, 7.931953944598931]}')])
env["res.partner"].search([("geo_polygon_field", "gis_intersects", {"type": "Point", "coordinates": [124.98706532650226, 7.931953944598931]})])
env["res.partner"].search([("geo_polygon_field", "gis_intersects", 'POINT (124.98706532650226 7.931953944598931)')])
env["res.partner"].search([("geo_polygon_field", "gis_intersects", Point(124.98706532650226, 7.931953944598931))])
```

### CODE REVIEW

File: spp_base_gis/models/gis_test_model.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+
+class GisTestModel(models.Model):
+    _name = "spp.base.gis.test.model"
+    _description = "GIS Test Model"
+
+    name = fields.Char()
+    geo_polygon_field = fields.GeoPolygonField()
+    geo_line = fields.GeoLineStringField()
+    geo_point = fields.GeoPointField()
```
Comment (jeremi) [inline spp_base_gis/models/gis_test_model.py]: It is better to move the test models into test projects. That way we do not install the test models in prod. Move the tests to this test project.

File: spp_base_gis/models/base.py [hunk 2]
```diff
@@ -93,3 +133,192 @@ def get_edit_info_for_gis_column(self, column):
...
+        constructed within the method. This query should be a valid PostGIS query that filters the
+        records in the database
+        :return: The method `raw_postgis_sql_query` is returning the result of the SQL query executed on
+        the database using the provided `postgis_query`. The method executes the query using
+        `self.env.cr.execute(query)` and then fetches all the results using `self.env.cr.fetchall()`.
+        The method will return a list of tuples where each tuple represents a row of the result set.
+        """
+        query = f"SELECT id FROM {self._table} WHERE {postgis_query}"
+        self.env.cr.execute(query)
+        return self.env.cr.fetchall()
```
Comment (jeremi) [inline spp_base_gis/models/base.py]: This seems risky with SQL Injection. Can we allow this with special right? By default, nobody should have it.

Comment (jeremi) [reply spp_base_gis/models/base.py]: Could we do a "safer" version that take a domain as parameter?

**Review by jeremi — CHANGES_REQUESTED** [review]

File: spp_base_gis/models/base.py (hunk 2 above)
Comment (reichie020212) [reply spp_base_gis/models/base.py]: noted on this. i'll work on this today

File: spp_base_gis/models/gis_test_model.py (hunk 1 above)
Comment (reichie020212) [reply spp_base_gis/models/gis_test_model.py]: Done.

File: spp_base_gis/models/base.py (hunk 2 above)
Comment (reichie020212) [reply spp_base_gis/models/base.py]: Done.

**Review by emjay0921 — APPROVED** [review]

**Review by gonzalesedwin1123 — APPROVED** [review]

**Review by emjay0921 — APPROVED** [review]

**Pull request merged**